lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,316 @@
+                      (intRangeValue.toVal_ &&
+                          (intRangeValue.toVal_.value() <= mem.toVal_)))));
+        });
+    if (selectCaseIntRangeValsIter == selectCaseIntRangeVals_.end()) {
+      selectCaseIntRangeVals_.push_back(intRangeValue);
+    } else {
+      ReportOverlapCaseValError(selectCaseIntRangeValsIter->source_, source);
+    }
+  }
+  void Insert(std::optional<std::string> lowerVal,
+      std::optional<std::string> upperVal, parser::CharBlock source) {
+    SelectCaseRangeType<std::string> stringRangeValue(
+        lowerVal, upperVal, source);
+    auto selectCaseStringValsIter = std::find_if(selectCaseStringVals_.begin(),
+        selectCaseStringVals_.end(), [&stringRangeValue](const auto &mem) {
+          return ((!stringRangeValue.fromVal_ ||
+                      ((stringRangeValue.fromVal_.value()).compare(mem.first) >=
+                          0)) &&
+              (!stringRangeValue.toVal_ ||
+                  ((stringRangeValue.toVal_.value()).compare(mem.first) <= 0)));
+        });
+    if (selectCaseStringValsIter != selectCaseStringVals_.end()) {
+      ReportOverlapCaseValError(selectCaseStringValsIter->second, source);
+    }
+    auto selectCaseStringRangeValsIter = std::find_if(
+        selectCaseStringRangeVals_.begin(), selectCaseStringRangeVals_.end(),
+        [&stringRangeValue](const auto &mem) {
+          return ((!mem.fromVal_ && !stringRangeValue.fromVal_) ||
+              (!mem.toVal_ && !stringRangeValue.toVal_) ||
+              ((!mem.fromVal_ ||
+                   (stringRangeValue.fromVal_ &&
+                       ((stringRangeValue.fromVal_.value())
+                               .compare(mem.fromVal_.value()) >= 0))) &&
+                  (!mem.toVal_ ||
+                      (stringRangeValue.fromVal_ &&
+                          ((stringRangeValue.fromVal_.value())
+                                  .compare(mem.toVal_.value()) <= 0)))) ||
+              ((!mem.fromVal_ ||
+                   (stringRangeValue.toVal_ &&
+                       (stringRangeValue.toVal_.value().compare(
+                            mem.fromVal_.value()) >= 0))) &&
+                  (!mem.toVal_ ||
+                      (stringRangeValue.toVal_ &&
+                          (stringRangeValue.toVal_.value().compare(
+                               mem.toVal_.value()) <= 0)))));
+        });
+    if (selectCaseStringRangeValsIter == selectCaseStringRangeVals_.end()) {
+      selectCaseStringRangeVals_.push_back(stringRangeValue);
+    } else {
+      ReportOverlapCaseValError(selectCaseStringRangeValsIter->source_, source);
+    }
+  }
+
+private:
+  template<typename T> struct SelectCaseRangeType {
+    SelectCaseRangeType(std::optional<T> fromVal, std::optional<T> toVal,
+        parser::CharBlock source)
+      : fromVal_(fromVal), toVal_(toVal), source_(source) {}
+    std::optional<T> fromVal_;
+    std::optional<T> toVal_;
+    parser::CharBlock source_;
```

Comment:
We only use the trailing underscore to mark private class data members. In this case `fromVal` etc. is fine.

Reply:
I have done this change.